fdbserver/StorageCache.actor.cpp

```diff
@@ -2012,32 +2077,283 @@ ACTOR Future<Void> watchInterface(StorageCacheData* self, StorageServerInterface
+				TEST( true ); // ShardWaitMetrics return case 2 (delayed)
+				req.reply.send( metrics );
+				break;
+			}
+		}
+
+		wait( delay(0) ); //prevent iterator invalidation of functions sending changes
+	}
+
+	auto rs = self->waitMetricsMap.modify( req.keys );
+	for(auto i = rs.begin(); i != rs.end(); ++i) {
+		auto &x = i->value();
+		for( int j = 0; j < x.size(); j++ ) {
+			if( x[j] == change ) {
+				swapAndPop(&x, j);
+				break;
+			}
+		}
+	}
+	self->waitMetricsMap.coalesce( req.keys );
+
+	if (error.code() != error_code_success ) {
+		if (error.code() != error_code_wrong_shard_server) throw error;
+		TEST( true );	// ShardWaitMetrics delayed wrong_shard_server()
+		req.reply.sendError(error);
+	}
+
+	return Void();
+}
+
+//Future<Void> StorageServerMetrics::waitMetrics(WaitMetricsRequest req, Future<Void> delay) {
+//	return ::scWaitMetrics(this, req, delay);
+//}
+
+ACTOR Future<Void> metricsCore( StorageCacheData* self, StorageServerInterface ssi ) {
+	state Future<Void> doPollMetrics = Void();
+
+	TraceEvent(SevDebug, "SCMetricsCoreAddingTraceCountersActor", self->thisServerID);
+	printf("\nSCMetricsCoreAddingTraceCountersActor\n");
+	printf("WaitMetricsMapSize: %d\n", self->metrics.waitMetricsMap.size());
+	self->actors.add(traceCounters("CacheMetrics", self->thisServerID, SERVER_KNOBS->STORAGE_LOGGING_DELAY, &self->counters.cc, self->thisServerID.toString() + "/CacheMetrics"));
+
+	loop {
+		choose {
+			when (WaitMetricsRequest req = waitNext(ssi.waitMetrics.getFuture())) {
+				if (!self->isReadable( req.keys )) {
+					TEST( true );	// waitMetrics immediate wrong_shard_server()
+					req.reply.sendError(wrong_shard_server());
+				} else {
+					TraceEvent(SevDebug, "SCMetricsCoreAddingWaitMetricsActor", self->thisServerID);
+					printf("\nSCMetricsCoreAddingWaitMetricsActor\n");
+					self->actors.add( self->metrics.waitMetrics( req, delayJittered( SERVER_KNOBS->STORAGE_METRIC_TIMEOUT ) ) );
+				}
+			}
+
+			when (SplitMetricsRequest req = waitNext(ssi.splitMetrics.getFuture())) {
+				if (!self->isReadable( req.keys )) {
+					TEST( true );	// splitMetrics immediate wrong_shard_server()
+					req.reply.sendError(wrong_shard_server());
+				} else {
+					self->metrics.splitMetrics( req );
+				}
+			}
+		
+			when (GetStorageMetricsRequest req = waitNext(ssi.getStorageMetrics.getFuture())) {
+				//StorageBytes sb = self->storage.getStorageBytes();
+				self->metrics.getStorageMetrics( req, self->counters.bytesInput.getRate(), self->versionLag);
+			}
+			when(ReadHotSubRangeRequest req = waitNext(ssi.getReadHotRanges.getFuture())) {
+				printf("\nSCGetReadHotRanges\n");
+				if (!self->isReadable(req.keys)) {
+					TEST(true); // readHotSubRanges immediate wrong_shard_server()
+					req.reply.sendError(wrong_shard_server());
+				} else {
+					self->metrics.getReadHotRanges(req);
+				}
+			}
+			when (wait(doPollMetrics) ) {
+				//printf("\nSCDOPollMetrics\n");
+				self->metrics.poll();
```

Comment:
Is this version the infinite polling version? It seems like `doPollMetrics` is always ready and thus poll will be always running.

Reply:
Yes, looks like I missed adding a delay here and that caused the actor to never yield. Adding a delay seems to have solved the infinite loop problem.